### PR TITLE
feat(stats): adding stats HTTP server to provide statistics about local p2pool

### PR DESCRIPTION
**Pull request by ksrichard**
Description
---

Adding statistics and expose them over an HTTP server.
Stats:
- [x] Pool hashrate
- [ ] Network hashrate (can be checked when tribes are implemented)
- [x] Number of miners
- [x] Last block won
- [x] Connected (to the network)
- [x] Connected time
- [x] estimated earning in XTM in 1min, 1h, 24h, 7d, 30d for all miners
- [x] estimated earning of pool in XTM in 1min, 1h, 24h, 7d, 30d
- [ ] Submitted/Accepted/Rejected blocks

Motivation and Context
---

Tari Universe wants to use metrics to show/use states about the current pool.

How Has This Been Tested?
---

- Running a miner and a p2pool node normally.
- Query http://127.0.0.1:19000/stats
- Checking output

Example output:
```json
{
  "connected": true,
  "connected_since": 1722519838,
  "num_of_miners": 2,
  "last_block_won": {
    "hash": "9c4fc0a3d996c6544e25b41ec0cf3a8bf771ac20fd0b7ae367e56ea597265d4f",
    "height": 5025,
    "timestamp": 1722518118,
    "miner_wallet_address": "f27mEvFXUZJCNFQM1MbkRwBnNJHwsK6JBxzrwVyYNm6bgGHpL1xDj7Gjevwj9caFrp23iLGUcysK6decdDL87sQCtL2"
  },
  "share_chain_height": 5196,
  "pool_hash_rate": [2536963],
  "pool_total_earnings": 27782378316,
  "pool_total_estimated_earnings": {
    "1min": 409165920,
    "1h": 24549955200,
    "1d": 589198924800,
    "1w": 4124392473600,
    "30d": 17675967744000
  },
  "total_earnings": {
    "f27mEvFXUZJCNFQM1MbkRwBnNJHwsK6JBxzrwVyYNm6bgGHpL1xDj7Gjevwj9caFrp23iLGUcysK6decdDL87sQCtL2": 5556473688,
    "f2CrtWaTZE3xWSxCkR1mR9Bszx321Ar2S1fPcBLNSNWiLBVwqkeankFjeTmdxYLuyeHg8oM4vSgsV1tjL4GSKEuy9pk": 8334710532
  },
  "estimated_earnings": {
    "f2CrtWaTZE3xWSxCkR1mR9Bszx321Ar2S1fPcBLNSNWiLBVwqkeankFjeTmdxYLuyeHg8oM4vSgsV1tjL4GSKEuy9pk": {
      "1min": 245499540,
      "1h": 14729972400,
      "1d": 353519337600,
      "1w": 2474635363200,
      "30d": 10605580128000
    },
    "f27mEvFXUZJCNFQM1MbkRwBnNJHwsK6JBxzrwVyYNm6bgGHpL1xDj7Gjevwj9caFrp23iLGUcysK6decdDL87sQCtL2": {
      "1min": 163666380,
      "1h": 9819982800,
      "1d": 235679587200,
      "1w": 1649757110400,
      "30d": 7070387616000
    }
  }
}
```

What process can a PR reviewer use to test or verify this change?
---

Do the steps at test.

Breaking Changes
---

- [x] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [ ] Other - Please specify

### CODE REVIEW

File: src/server/http/stats/handlers.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    // connected
+    let connected = state.peer_store.peer_count().await > 0;
+
+    // collect number of miners
+    let num_of_miners = chain
+        .iter()
+        .map(|block| block.miner_wallet_address())
+        .filter(|addr_opt| addr_opt.is_some())
+        .map(|addr| addr.as_ref().unwrap().to_base58())
+        .unique()
```
Comment (therealdannzor) [inline src/server/http/stats/handlers.rs:37]: q: when do we have duplicate addresses? if someone who previous left the pool re-joins?

File: src/sharechain/in_memory.rs [hunk 2]
```diff
@@ -341,4 +338,51 @@ impl ShareChain for InMemoryShareChain {
...
+            .iter()
+            .flat_map(|level| level.blocks.clone())
+            .sorted_by(|block1, block2| block1.timestamp().cmp(&block2.timestamp()))
+            .tail(BLOCKS_WINDOW);
+
+        // calculate average block time
+        let blocks = blocks.collect_vec();
+        let mut block_times_sum = 0;
+        let mut block_times_count: u64 = 0;
+        for i in 0..blocks.len() {
```
Comment (therealdannzor) [inline src/sharechain/in_memory.rs:358]: would it work to only take the first and last block to get the time passed, divided by amount of blocks?

**Review by therealdannzor — APPROVED** [review]
Looks good to me, just two minor questions :)

File: src/server/http/stats/handlers.rs (hunk 1 above)
Comment (ksrichard) [reply src/server/http/stats/handlers.rs:37]: ah no, we are iterating over the current strongest chain, so as a miner could mine many blocks, then after mapping to wallet address we definitely will have duplicates, so thatswhy I needed `.unique()`

File: src/sharechain/in_memory.rs (hunk 2 above)
Comment (ksrichard) [reply src/sharechain/in_memory.rs:358]: I don't think so as here I'm interested in the average time between 2 blocks, I feel like its maybe more precise

**Review by sdbondi — APPROVED** [review]

**Pull request merged**